client/mm/libxc/binance.go

```diff
@@ -0,0 +1,1168 @@
+		updater <- struct{}{}
+	}
+}
+
+func (bnc *binance) handleOutboundAccountPosition(update *bncStreamUpdate) {
+	bnc.log.Tracef("received outboundAccountPosition: %+v", update)
+
+	bnc.balanceMtx.Lock()
+	for _, bal := range update.Balances {
+		symbol := strings.ToLower(bal.Asset)
+		bnc.balances[symbol] = &bncBalance{
+			available: bal.Free,
+			locked:    bal.Locked,
+		}
+	}
+	bnc.balanceMtx.Unlock()
+	bnc.sendCexUpdateNotes()
+}
+
+func (bnc *binance) getTradeUpdater(orderID string) (chan *TradeUpdate, error) {
+	bnc.tradeUpdaterMtx.RLock()
+	defer bnc.tradeUpdaterMtx.RUnlock()
+
+	updaterID, found := bnc.tradeToUpdater[orderID]
+	if !found {
+		return nil, fmt.Errorf("updater not found for trade ID %v", orderID)
+	}
+	updater, found := bnc.tradeUpdaters[updaterID]
+	if !found {
+		return nil, fmt.Errorf("no updater with ID %v", updaterID)
+	}
+
+	return updater, nil
+}
+
+func (bnc *binance) handleExecutionReport(update *bncStreamUpdate) {
+	bnc.log.Tracef("received executionReport: %+v", update)
+
+	updater, err := bnc.getTradeUpdater(update.ClientOrderID)
+	if err != nil {
+		bnc.log.Errorf("error getting trade updater: %v", err)
+		return
+	}
+
+	status := update.CurrentOrderStatus
+
+	updater <- &TradeUpdate{
+		TradeID:  update.ClientOrderID,
+		Complete: status == "FILLED" || status == "CANCELED" || status == "REJECTED" || status == "EXPIRED",
+	}
+}
+
+func (bnc *binance) getUserDataStream() (err error) {
+	streamHandler := func(b []byte) {
+		u, err := decodeStreamUpdate(b)
+		if err != nil {
+			bnc.log.Errorf("error unmarshaling user stream update: %v", err)
+			bnc.log.Errorf("raw message: %s", string(b))
+			return
+		}
+		switch u.EventType {
+		case "outboundAccountPosition":
+			bnc.handleOutboundAccountPosition(u)
+		case "executionReport":
+			bnc.handleExecutionReport(u)
+		case "balanceUpdate":
+			// TODO: check if we need this.. is outbound account position enough?
+			bnc.log.Tracef("received balanceUpdate: %s", string(b))
+		}
+	}
+
+	var listenKey string
+	newConn := func() (*dex.ConnectionMaster, error) {
+		listenKey, err = bnc.getListenID()
+		if err != nil {
+			return nil, err
+		}
+
+		hdr := make(http.Header)
+		hdr.Set("X-MBX-APIKEY", bnc.apiKey)
+		conn, err := comms.NewWsConn(&comms.WsCfg{
+			URL: bnc.wsURL + "/ws/" + listenKey,
+
+			// Is this necessary for user data stream??
+			PingWait: time.Minute * 4,
+
+			ReconnectSync: func() {
+				bnc.log.Debugf("Binance reconnected")
+			},
+			ConnectEventFunc: func(cs comms.ConnectionStatus) {},
+			Logger:           bnc.log.SubLogger("BNCWS"),
+			RawHandler:       streamHandler,
+			ConnectHeaders:   hdr,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		cm := dex.NewConnectionMaster(conn)
+		if err = cm.ConnectOnce(bnc.ctx); err != nil {
+			return nil, fmt.Errorf("websocketHandler remote connect: %v", err)
+		}
+
+		return cm, nil
+	}
+
+	cm, err := newConn()
+	if err != nil {
+		return fmt.Errorf("error initializing connection: %v", err)
+	}
```

Comment:
I'm thinking it may be better if the initial connection fails to not even start the market maker, because this will be more obvious to the user and they may have to fix something. What do you think? But it's good to bring this point up.. there should be notifications to the user if later the connection goes down and cannot reconnect.

Reply:
I would assume they would notice the bots not connecting... but then again this is started from the UI bot menu? I guess it would be pretty weird for binance to be down. Anything that's not a connection error, like a bad api key?, should be fatal, as that will never be ok. So maybe don't change.

But yeah notification notes to the UI would be great on errors here, if this is started in the UI.